src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -405,7 +405,14 @@ protected void InitializeReferences()
             try
             {
                 // Do the scan starting at the requested position
-                Match? match = runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                Match? match = runner.runmatch;
+                runner.runmatch = null; // Reset runmatch after saving a copy.
+
+                if (match != null && !quick)
+                {
+                    match.Text = input.ToString(); // Only box in case a match Object is going to be returned to the user.
+                }
```

Comment:
With a single Scan virtual implementations override, we're going to need to change the separate `Run<TState>` I'd added below as it won't be able to call its own special callback-invoking Scan loop.  The cleanest option is likely to just build Split and Replace around whatever Enumerate method we end up exposing, and just get rid of the below `Run<TState>` entirely.